File: remember/separators.py
import numpy as np

from remember.config import Config
from remember.model import Rectangle


def extract(config: Config, matrix: np.ndarray, *, axis: int) -> list[int]:
    ys = []
    current_ys = []
    # noinspection PyUnresolvedReferences
    for y, is_separator in enumerate(
        ((matrix <= config.black_threshold).mean(axis=1 - axis) > config.separator_threshold).tolist() + [False]
    ):
        if is_separator:
            current_ys.append(y)
        elif len(current_ys) > 0:
            ys.append(np.round(np.mean(current_ys)).astype(int))
            current_ys = []
    return ys


def extract_from_box(config: Config, matrix: np.ndarray, box: Rectangle, *, axis: int) -> list[int]:
    return extract(config, matrix[box.y1 : box.y2, box.x1 : box.x2], axis=axis)

File: src/separators.js
"use strict";

if (!window.reMember) {
    window.reMember = {};
}

window.reMember.separators = {
    extract(config, image, transpose) {
        const matrix = transpose ? reMember.util.transpose(image.bw_data) : image.bw_data;
        const ys = [];
        let current_ys = [];
        for (let y = 0; y <= matrix.length; y++) {
            const is_separator = y < matrix.length && matrix[y].filter(c => c <= config.black_threshold).length / matrix[y].length > config.separator_threshold;
            if (is_separator) {
                current_ys.push(y);
            } else if (current_ys.length > 0) {
                ys.push(Math.round(current_ys.reduce((a, b) => a + b) / current_ys.length));
                current_ys = [];
            }
        }
        return ys;
    },
    extract_from_box(config, image, box, transpose) {
        return reMember.separators.extract(config, image.crop(box), transpose);
    }
};
